WikiWatershed/model-my-watershed
# Scaffold Excel Generation Endpoint

## Overview

This scaffolds two new endpoints:

- `/mmw/modeling/worksheet/`: This endpoint takes a shape, pairs it with HUC-12s, and in the future will start a Celery job that calculates a number of values for each pair, resulting in a JSON blob
- `/export/worksheet/`: This endpoint takes the JSON blob generated above and returns a ZIP file containing Excel worksheets corresponding to each pair

Currently, the first endpoint returns an empty job, and prints out the pairs in the debug log. The second endpoint takes an array of objects, and returns a ZIP of Excel worksheets that are hardcoded to have the same value, for as many objects were in the input array. This scaffolding will be expanded upon in future work.

Next steps:

- Create a Celery queue that generates the JSON blob
- Update the export endpoint to generate files corresponding to the HUC-12 names, rather than `0.xlsx` like it currently does
- Add input validation
- Add tests?

See commit messages for details.

Connects #3185 

### Demo

Getting a Celery Job response shape from the modeling endpoint where a shape is submitted:

```http
$ http :8000/api/modeling/worksheet/ Authorization:"Token $MMW_API_TOKEN" < triple-hucs.geojson

HTTP/1.1 200 OK
Allow: POST, OPTIONS
Connection: keep-alive
Content-Encoding: gzip
Content-Type: application/json
Date: Tue, 29 Oct 2019 20:50:32 GMT
Server: nginx
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Accept, Cookie, Origin

{
    "job": "00000000-0000-0000-0000-000000000000",
    "status": "started"
}
```

The clipped shapes are currently printed in the `debugserver` output. Porting them to GeoJSON.io, we can visualize the results like this:

![image](https://user-images.githubusercontent.com/1430060/67607643-ee44a100-f720-11e9-96c2-db138de8eaf8.png)

![image](https://user-images.githubusercontent.com/1430060/67607708-23e98a00-f721-11e9-90e7-6f0ff355c66e.png)

Getting a ZIP file containing Excel Spreadsheets back from the export endpoint:

```http
$ http --download :8000/export/worksheet/ < test.json

HTTP/1.1 200 OK
Allow: POST, OPTIONS
Connection: keep-alive
Content-Disposition: attachment; filename="MMW_BMP_Spreadsheets.zip"
Content-Length: 244138
Content-Type: application/zip
Date: Tue, 29 Oct 2019 20:31:05 GMT
Server: nginx
Vary: Accept, Cookie

Downloading 238.42 kB to "MMW_BMP_Spreadsheets.zip"
Done. 238.42 kB in 0.00400s (58.15 MB/s)
```

Excel file with Stream Lengths filled out:

![image (2)](https://user-images.githubusercontent.com/1430060/67825484-d6ae4500-fa9f-11e9-9c68-8fae8560a4c9.png)

## Testing Instructions

* Checkout this branch and run the server `$ ./scripts/debugserver.sh`
* Go to [:8000/](http://localhost:8000/) and turn on the HUC-12 boundary layer in the layer selector
* Draw a shape that intersects 2 or more HUC-12s
* Download a GeoJSON of the shape
* Using your user's MMW API Token, fire off a request to the `/mmw/modeling/worksheet/` endpoint with the GeoJSON of the shape as the body
  - [x] Ensure you see a Celery-like output
  - [x] Ensure you see the cut shapes paired with their HUC-12s in the `debugserver` output
* Fire off a request to the `/export/worksheet/` endpoint, with an array of objects like `[{},{}]` as the body
  - [x] Ensure you get a ZIP file in response
  - [x] Ensure it has as many Excel worksheets as objects in the input array
  - [x] Ensure the Excel worksheets have values in the stream length section of the MMW Output tab
